juju/juju
# Payload hook tools move

## Description of change

Payload hook tools were implemented under a component based architecture which made them very different to the rest of the code. This cased discrepancies in behavior, maintainability issues, etc. 

This PR will need to be followed up by an addition of payload validation against unit charm at state layer. This was not added as the change was introducing friction in other areas like model migration. This change will need to be done separately to minimize surface difference of this move.

As payload hook logic had to be moved into uniter, uniter's facades needed a version bump.

## QA steps

Deploy charm with payloads, track and untrack them, as well as set status to a different value - there should not be any differences in current behavior (except for the lack of validation as previously stated).

## Documentation changes

Most of the documentation does not need to change as this is an internal change.
However, 'juju hook-tool' now should also include payload hook tools.

$ juju help-tool
action-fail              set action fail status with message
action-get               get action parameters
action-set               set action results
add-metric               add metrics
application-version-set  specify which version of the application is deployed
close-port               ensure a port or range is always closed
config-get               print application configuration
is-leader                print application leadership status
juju-log                 write a message to the juju log
juju-reboot              Reboot the host machine
leader-get               print application leadership settings
leader-set               write application leadership settings
network-get              get network config
open-port                register a port or range to open
opened-ports             lists all ports or ranges opened by the unit
payload-register         register a charm payload with juju
payload-status-set       update the status of a payload
payload-unregister       stop tracking a payload
relation-get             get relation settings
relation-ids             list all relation ids with the given relation name
relation-list            list relation units
relation-set             set relation settings
resource-get             get the path to the locally cached resource file
status-get               print status information
status-set               set status information
storage-add              add storage instances
storage-get              print information for storage instance with specified id
storage-list             list storage attached to the unit
unit-get                 print public-address or private-address


## Bug reference

https://bugs.launchpad.net/juju/+bug/1628919
